File: ModelRelief/Features/Meshes/Mesh.cs
// -----------------------------------------------------------------------
// <copyright file="Mesh.cs" company="ModelRelief">
// Copyright (c) ModelRelief. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace ModelRelief.Dto
{
    using System;
    using System.ComponentModel.DataAnnotations;
    using AutoMapper;
    using FluentValidation;
    using Microsoft.AspNetCore.Http;
    using ModelRelief.Domain;

    using Newtonsoft.Json;

    /// <summary>
    /// Represents a DataTransferObject (DTO) for a Mesh.
    /// </summary>
    public class Mesh : IGeneratedFileModel, IProjectModel
    {
        public int Id { get; set; }

        [Display(Name = "Mesh Name")]
        public string Name { get; set; }
        public string Description { get; set; }

        public MeshFormat Format { get; set; }

        [IgnoreMap]
        [JsonIgnore]
        [Display(Name = "File (Upload)")]
        public IFormFile FormFile { get; set; }

        // Navigation Properties
        public int? ProjectId { get; set; }
        public Project Project { get; set; }

        public int? CameraId { get; set; }
        public Camera Camera { get; set; }

        public int? DepthBufferId { get; set; }
        public DepthBuffer DepthBuffer { get; set; }

        public int? NormalMapId { get; set; }
        public NormalMap NormalMap { get; set; }

        public int? MeshTransformId { get; set; }
        public MeshTransform MeshTransform { get; set; }

        // not exposed in UX; API only
        public DateTime? FileTimeStamp { get; set; }
        public bool FileIsSynchronized { get; set; }

        /// <summary>
        /// Initializes a new instance of the <see cref="Mesh"/> class.
        /// Constructor.
        /// </summary>
        public Mesh()
        {
        }
    }

    /// <summary>
    /// FV validator to support Views and model-binding validation.
    /// </summary>
    public class MeshValidator : AbstractValidator<Mesh>
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="MeshValidator"/> class.
        /// Constructor.
        /// </summary>
        public MeshValidator()
        {
            RuleFor(m => m.Name)
                .NotNull().WithMessage("The Name property is required.");

            RuleFor(m => m.Description)
                .NotNull().WithMessage("The Description property is required.")
                .MinimumLength(3).WithMessage("The Description must be three or more characters.");

            RuleFor(m => m.Format)
                .NotEmpty().WithMessage("The file format must be provided.");
        }
    }

    /// <summary>
    /// AutoMapper mapping profile.
    /// </summary>
    public class MeshMappingProfile : Profile
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="MeshMappingProfile"/> class.
        /// Constructor.
        /// </summary>
        public MeshMappingProfile()
        {
            CreateMap<Domain.Mesh, Mesh>().ReverseMap();
        }
    }
}

File: ModelRelief/Scripts/Models/Mesh/Mesh.ts
// ------------------------------------------------------------------------//
// ModelRelief                                                             //
//                                                                         //
// Copyright (c) <2017-2018> Steve Knipmeyer                               //
// ------------------------------------------------------------------------//
"use strict";

import * as Dto from "Scripts/Api/V1//Models/DtoModels";
import * as THREE from "three";

import {IGeneratedFileModel} from "Scripts/Api/V1/Interfaces/IGeneratedFileModel";
import {MeshFormat} from "Scripts/Api/V1/Interfaces/IMesh";
import {GeneratedFileModel} from "Scripts/Api/V1/Models/GeneratedFileModel";
import {BaseCamera} from "Scripts/Models/Camera/Camera";
import {CameraFactory} from "Scripts/Models/Camera/CameraFactory";
import {DepthBuffer} from "Scripts/Models/DepthBuffer/DepthBuffer";
import {MeshTransform} from "Scripts/Models/MeshTransform/MeshTransform";
import {Project} from "Scripts/Models/Project/Project";

/**
 * @description Represents a mesh.
 * @export
 * @class Mesh
 * @extends {GeneratedFileModel}
 */
export class Mesh extends GeneratedFileModel {

    //#region Properties
    /**
     * @description Returns the width.
     * @readonly
     * @type {number}
     */
    get width(): number {
        return this.depthBuffer.width;
    }

    /**
     * @description Returns the height.
     * @readonly
     * @type {number}
     */
    get height(): number {
        return this.depthBuffer.height;
    }

    /**
     * @description Returns a Mesh instance through an HTTP query of the Id.
     * @static
     * @param {number} id Mesh Id.
     * @returns {Promise<Mesh>}
     */
    public static async fromIdAsync(id: number ): Promise<Mesh> {

        if (!id)
            return undefined;

        const mesh = new Dto.Mesh ({
            id,
        });
        const meshModel = await mesh.getAsync();
        return Mesh.fromDtoModelAsync(meshModel);
    }

    /**
     * @description Constructs an instance from a DTO model.
     * @returns {Mesh}
     */
    public static async fromDtoModelAsync(dtoMesh: Dto.Mesh): Promise<Mesh> {

        // constructor
        const mesh = new Mesh ({
            id          : dtoMesh.id,
            name        : dtoMesh.name,
            description : dtoMesh.description,
            },
        );

        mesh.fileTimeStamp      = dtoMesh.fileTimeStamp;
        mesh.fileIsSynchronized = dtoMesh.fileIsSynchronized;
        mesh.format             = dtoMesh.format;

        mesh.project = await Project.fromIdAsync(dtoMesh.projectId);
        mesh.camera  = await CameraFactory.ConstructFromIdAsync(dtoMesh.cameraId);
        mesh.depthBuffer    = await DepthBuffer.fromIdAsync(dtoMesh.depthBufferId);
        mesh.meshTransform  = await MeshTransform.fromIdAsync(dtoMesh.meshTransformId);

        return mesh;
    }

    public format: MeshFormat;

    // Navigation Properties
    public project: Project;
    public camera: BaseCamera;
    public depthBuffer: DepthBuffer;
    public meshTransform: MeshTransform;

    /**
     * Creates an instance of Mesh.
     * @param {IGeneratedFileModel} [parameters={}] GeneratedFileModel properties.
     */
    constructor(parameters: IGeneratedFileModel) {

        parameters.name        = parameters.name        || "Mesh";
        parameters.description = parameters.description || "Mesh";

        super(parameters);

        this.initialize();
    }

    /**
     * @description Perform setup and initialization.
     */
    public initialize(): void {
    }

    /**
     * @description Returns a DTO Mesh from the instance.
     * @returns {Dto.Mesh}
     */
    public toDtoModel(): Dto.Mesh {

        const mesh = new Dto.Mesh({
            id              : this.id,
            name            : this.name,
            description     : this.description,

            format          : this.format,

            projectId       : this.project ? this.project.id : undefined,
            cameraId        : this.camera ? this.camera.id : undefined,
            depthBufferId   : this.depthBuffer ? this.depthBuffer.id : undefined,
            meshTransformId : this.meshTransform ? this.meshTransform.id : undefined,

            fileTimeStamp      : this.fileTimeStamp,
            fileIsSynchronized : this.fileIsSynchronized,
        });

        return mesh;
    }
    //#endregion
}
